File: proj4a/Merge-Sort/merge.pl
:- use_module(library(random)).
:- use_module(library(statistics)).

% Gera uma lista de números aleatórios de tamanho N com elementos entre 0 e MaxValue
generate_random_list(N, MaxValue, List) :-
    length(List, N),
    maplist(random_between(0, MaxValue), List).

% Gera 10 listas aleatórias do mesmo tamanho
generate_random_lists(10, Length, MaxValue, Lists) :-
    length(Lists, 10),
    maplist(generate_random_list(Length, MaxValue), Lists).

% Merge Sort
merge_sort([], []).
merge_sort([X], [X]).
merge_sort(List, Sorted) :-
    List = [_, _|_],  % ensures List has at least two elements
    divide(List, L1, L2),
    merge_sort(L1, Sorted1),
    merge_sort(L2, Sorted2),
    merge(Sorted1, Sorted2, Sorted).

divide(List, L1, L2) :-
    length(List, N),
    Half is N // 2,
    length(L1, Half),
    append(L1, L2, List).

merge([], L, L).
merge(L, [], L).
merge([X|T1], [Y|T2], [X|T]) :-
    X =< Y,
    merge(T1, [Y|T2], T).
merge([X|T1], [Y|T2], [Y|T]) :-
    X > Y,
    merge([X|T1], T2, T).

% Ordena e mede o tempo de execução de uma lista
sort_and_time_list(List) :-
    statistics(walltime, [_ | [_]]),
    merge_sort(List, _Sorted),
    statistics(walltime, [_ | [ExecutionTime]]),
    format('Tempo de execução para lista: ~3f segundos~n', [ExecutionTime / 1000.0]).

% Predicado principal para gerar e ordenar as listas
main :-
    % Definir o tamanho e valor máximo dos elementos das listas
    Length = 1000,
    MaxValue = 100,

    % Gerar 10 listas aleatórias
    generate_random_lists(10, Length, MaxValue, Lists),

    % Ordenar e medir o tempo de execução de cada lista
    maplist(sort_and_time_list, Lists).

File: proj4a/Merge-Sort/merge.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void generate_random_list(int list[], int size, int max_value) {
    for (int i = 0; i < size; i++) {
        list[i] = rand() % (max_value + 1);
    }
}

void merge(int list[], int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;
    int leftArr[n1], rightArr[n2];
    
    for (int i = 0; i < n1; i++)
        leftArr[i] = list[left + i];
    for (int j = 0; j < n2; j++)
        rightArr[j] = list[mid + 1 + j];
    
    int i = 0, j = 0, k = left;
    while (i < n1 && j < n2) {
        if (leftArr[i] <= rightArr[j]) {
            list[k] = leftArr[i];
            i++;
        } else {
            list[k] = rightArr[j];
            j++;
        }
        k++;
    }
    
    while (i < n1) {
        list[k] = leftArr[i];
        i++;
        k++;
    }
    
    while (j < n2) {
        list[k] = rightArr[j];
        j++;
        k++;
    }
}

void merge_sort(int list[], int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        
        merge_sort(list, left, mid);
        merge_sort(list, mid + 1, right);
        
        merge(list, left, mid, right);
    }
}

int main() {
    srand(time(NULL));
    long int size = 100000;

    int lists[10][size];

    // Gerar 10 listas aleatórias
    for (int i = 0; i < 10; i++) {
        generate_random_list(lists[i], size, 100);
    }

    // Ordenar cada lista usando Merge Sort e medir o tempo de execução para cada uma
    for (int i = 0; i < 10; i++) {
        clock_t start_time = clock();
        merge_sort(lists[i], 0, size - 1);
        clock_t end_time = clock();
        double execution_time = ((double) (end_time - start_time)) / CLOCKS_PER_SEC;
        printf("Tempo de execução para a lista %d: %f \n", i + 1, execution_time);
    }

    return 0;
}
